fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/impl/ResourceReferenceDAO.java

```diff
@@ -756,78 +753,104 @@ public void upsertCommonTokenValues(List<ResourceTokenValueRec> values) {
+                if (inList.length() > 0) {
+                    inList.append(",");
+                }
+                inList.append("(?,?)");
+            }
+
+            final String paramListStr = paramList.toString();
+            doCommonTokenValuesUpsert(paramListStr, sortedTokenValuesSub);
+
+            // Now grab the ids for the rows we just created. If we had a RETURNING implementation
+            // which worked reliably across all our database platforms, we wouldn't need this
+            // second query.
+            // Derby doesn't support IN LISTS with multiple members, so we have to join against
+            // a VALUES again. No big deal...probably similar amount of work for the database
+            StringBuilder select = new StringBuilder();
+            select.append("     SELECT ctv.code_system_id, ctv.token_value, ctv.common_token_value_id FROM ");
+            select.append("     (VALUES ").append(paramListStr).append(" ) AS v(token_value, code_system_id) ");
+            select.append("       JOIN common_token_values ctv ");
+            select.append("              ON ctv.token_value = v.token_value ");
+            select.append("             AND ctv.code_system_id = v.code_system_id ");
+
+            // Grab the ids
+            Map<CommonTokenValue, Long> idMap = new HashMap<>();
+            try (PreparedStatement ps = connection.prepareStatement(select.toString())) {
+                int a = 1;
+                for (CommonTokenValue tv: sortedTokenValuesSub) {
+                    ps.setString(a++, tv.getTokenValue());
+                    ps.setInt(a++, tv.getCodeSystemId());
+                }
+
+                ResultSet rs = ps.executeQuery();
+                while (rs.next()) {
+                    // SELECT code_system_id, token_value...note codeSystem not required
+                    CommonTokenValue key = new CommonTokenValue(null, rs.getInt(1), rs.getString(2));
+                    idMap.put(key, rs.getLong(3));
+                }
+            } catch (SQLException x) {
+                throw translator.translate(x);
+            }
+
+            // Now update the ids for all the matching systems in our list
+            for (ResourceTokenValueRec xr: values) {
+                // ignore entries with null tokenValue elements - we don't store them in common_token_values
+                if (xr.getTokenValue() != null) {
+                    CommonTokenValue key = new CommonTokenValue(null, xr.getCodeSystemValueId(), xr.getTokenValue());
+                    Long id = idMap.get(key);
+                    if (id != null) {
+                        xr.setCommonTokenValueId(id);
+
+                        // update the thread-local cache with this id. The values aren't committed to the shared cache
+                        // until the transaction commits
+                        cache.addTokenValue(key, id);
+                    }
                 }
             }
+            low = high + 1;
```

Comment:
moves the list forward

Reply:
with subList the high is exclusive, so I think you're skipping an entry here. Should be `low = high` I believe.

Reply:
I think you are right... let me double check.  I know I added low/high logging just to verify this.

Reply:
changed the implementation to use a single loop